### DISCUSSION TITLE
Import inventory from and export to csv file

### DISCUSSION THREAD

**Discussion by aairfccha** [radioactivedecay/radioactivedecay#94] [Show and tell]
For practical use of this package, it can be somewhat annoying to type inventory data in manually and writing it into a python script is only a limited improvement compared to importing an inventory from a table in a common format and also write the result of decay calculations in a new file. This code (in no small part hacked together from sample code for csv import and export) does just that: it imports an inventory from a |-delimited csv-file, decays the inventory for a set period (here 15 years) and writes the result into a new csv-file. It should require no additional packages (csv is part of the standard library) and for a quick sanity check of in- and output, it also has some easily human-readable output to the console.

Here is a sample inventory, and yes you can arbitrarily mix and match units supported by radioactivedecay. However there is no sanity check, the way the actual import is done, nuclides which appear more than once are merrily added together.

```
H-1|250|g
H-2|0.005|g
H-3|0.1337|g
He-3|0.001329|g
He-4|8467|g
Ra-226|1|Ci
Pu-239|20|g
Po-210|20|g
```

Now for the actual code.

```
import radioactivedecay as rd
import csv

inputfile = 'Import_Test.csv'
outputfile = 'Export_test.csv'
decaytime_1 = 15
decaytime_1_unit = 'y'
mass_unit = 'g'

with open(inputfile, newline='') as csvfile:
    importtest = csv.reader(csvfile, delimiter='|')
    line = 0
    print('initial inventory as imported')
    for row in importtest:
        print(row[1],row[2],'of',row[0])
        if line == 0:
            inv_t0 = rd.Inventory({row[0]: float(row[1])}, row[2])
        else:
            inv_t0.add({row[0]: float(row[1])}, row[2])
        line += 1
        
inv_t1 = inv_t0.decay(decaytime_1, decaytime_1_unit)

#required (?) to turn output as dictionary into actual dictionary
inv_t0_mass = inv_t0.masses(mass_unit)
inv_t1_mass = inv_t1.masses(mass_unit)

print()
print('initial inventory,', mass_unit)
for x in range(0, len(inv_t0_mass)):
    index = list(inv_t0_mass)[x]
    print(inv_t0_mass[index], mass_unit, 'of', index)

print()
print('inventory after', str(decaytime_1), decaytime_1_unit, 'of decay,', mass_unit)
for x in range(0, len(inv_t1_mass)):
    index = list(inv_t1_mass)[x]
    print(inv_t1_mass[index], mass_unit, 'of', index)

#maybe comment this block out to minimise writes to storage when testing
with open(outputfile, 'w', newline='') as csvfile:
    spamwriter = csv.writer(csvfile, delimiter='|')
    for x in range (0, len(inv_t1_mass)):
        index = list(inv_t1_mass)[x]
        spamwriter.writerow([str(index), str(inv_t1_mass[index]), mass_unit])
```

**Comment by alexmalins** [comment]
Hey @aairfccha - this is really great, thanks for sharing your code 🙌 

If it is ok, I'll use this as a basis for adding a `rd.read_csv()` function - I can imagine reading a csv like this with nuclides with different units is something a lot of people will find useful.

Likewise, also thinking of doing a `.to_csv()` method for the inventory classes to write the contents out to csv (in the chosen units, delimiter etc.)  🚀

**Comment by alexmalins** [comment]
Hey @aairfccha - I just released v0.4.20 which does the first part. It has a new `rd.read_csv()` function for creating inventories from a CSV file. See the docs for how to use it: end of this [section](https://radioactivedecay.github.io/tutorial.html#creating-inventories-of-nuclides) and in the API [docs](https://radioactivedecay.github.io/tutorial.html#creating-inventories-of-nuclides).

At some point I'll add a `.to_csv()` method for writing decay calculation results (i.e. content of an Inventory) to a CSV file.

**Reply by alexmalins** [reply]
Release 0.4.21 just added a `.to_csv()` method for exporting inventories to CSV files 🚀

**Reply by aairfccha** [reply]
Nice to see that my tinkering inspired an addition to the official code, although I don't really see the point of not writing the units into the output file. I guess you can write it in the table header, but no unit column makes the files generated by the export function incompatible with the import function (or is Bq assumed as default?). Maybe add an optional `assume_unit` parameter to `.read_csv()` to allow reliably importing those files as well (with a unit column overriding the assumed unit if present)?

And now that quite substantial inventories can easily be imported, the topics of decay heat and missing nuclides might become more relevant to prospective users, since the result of SNF burn-up calculations can now be imported with relative ease (in fact, doing this was the reason I wrote this code in the first place): Those inventories both include a big zoo of weird nuclides and the decay heat is often somewhat relevant.

**Reply by alexmalins** [reply]
hi @aairfccha 

> although I don't really see the point of not writing the units into the output file. I guess you can write it in the table header

yes, the `.to_csv()` [method](https://radioactivedecay.github.io/inventory.html#radioactivedecay.inventory.Inventory.to_csv) supports either writing a units column or not:

```
# write no units column, puts unit in the header:
inv.to_csv("no_units_col.csv", units="GBq", delimiter=",", write_units=False, header=["nuclide", "activity (GBq)"])

# write units column:
inv.to_csv("with_units_col.csv", units="GBq", delimiter=",", write_units=True, header=["nuclide", "activity", "unit"])
```


>  but no unit column makes the files generated by the export function incompatible with the import function (or is Bq assumed as default?). Maybe add an optional assume_unit parameter to .read_csv() to allow reliably importing those files as well (with a unit column overriding the assumed unit if present)?

The import [function](https://radioactivedecay.github.io/fileio.html#radioactivedecay.fileio.read_csv) `read_csv()` supports either two or three column CSV files, so it doesn't make it incompatible. I think the existing `read_csv()` `units` parameter already does your suggestion if I understood it correctly:

```
# assumes units of Bq as nothing specified and no units col:
rd.read_csv("no_units_col.csv", delimiter=",", skip_rows=1)  # skip_rows=1 needed to ignore header row

# specify correct units of GBq:
rd.read_csv("no_units_col.csv", units="GBq", delimiter=",", skip_rows=1)  

# reads units from third col in file, these override specified units:
rd.read_csv("units_col.csv", units="Ci", delimiter=",", skip_rows=1)
```


>And now that quite substantial inventories can easily be imported, the topics of decay heat and missing nuclides might become more relevant to prospective users, since the result of SNF burn-up calculations can now be imported with relative ease (in fact, doing this was the reason I wrote this code in the first place): Those inventories both include a big zoo of weird nuclides and the decay heat is often somewhat relevant.

Yes thank you, I hope to find the time someday to be able to put more work into this package.

**Comment by aairfccha** [comment]
With those import and export functions, the original script can be obviously a lot shorter and a few nifty things are possible. For example, take the test inventory above (with a slight structural update),
```
nuclide|quantity|units
H-1|267|g
H-2|0.004995|g
H-3|0.1337|g
He-3|0.001329|g
He-4|8467|g
Ra-226|1|Ci
Pu-239|20|g
Po-210|20|g
```
save it as `Test_t0.csv` and feed it to this code:
```
import radioactivedecay as rd

decaytime = 5
decaytime_unit = 'y'
export_unit = 'g'
inputfile = 'Test_t0.csv'
outputfile = 'Test_'+'t'+str(decaytime)+decaytime_unit+'_'+str(export_unit)+'.csv'

# Read inventory from specified input file (assuming header and units column)
inv_t0 = rd.read_csv(inputfile, delimiter='|', skip_rows=1)

# decay inventory by specified time
inv_t1 = inv_t0.decay(decaytime, decaytime_unit)

# writes decayed inventory to specified output file (with header and units column)
inv_t1.to_csv(outputfile, units=export_unit, delimiter='|', write_units=True, header=["nuclide", "quantity", "units"])

print('done')
```
This script automagically writes both the decay time and the unit of data export into the file name to avoid overwriting old files when you want to calculate both masses and activities or inventories after a bunch of different decay times. Writing decay time into the file name would have been possible from the beginning, but the output units were much less flexible before.

Playing around a little opened the next issue: High precision calculations. As I understand the tutorial, they use the same decay routine fed with a different type of inventory, so apparently I need to do a conversion after import. Is there a more elegant solution than this kludge of writing the inventory into a dictionary and reading it back as a new high precision inventory?
```
inv_t0_mol = inv_t0.moles()
high_prec_t0 = rd.InventoryHP(inv_t0_mol, 'mol')
```
